File: src/mrtjp/projectred/illumination/RenderHalo.java
package mrtjp.projectred.illumination;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import mrtjp.projectred.core.Configurator;
import mrtjp.projectred.core.libmc.PRColors;
import net.minecraft.client.multiplayer.WorldClient;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.world.World;
import net.minecraftforge.client.event.RenderWorldLastEvent;
import net.minecraftforge.event.ForgeSubscribe;

import org.lwjgl.opengl.GL11;

import codechicken.lib.render.CCRenderState;
import codechicken.lib.render.RenderUtils;
import codechicken.lib.vec.BlockCoord;
import codechicken.lib.vec.Cuboid6;
import codechicken.lib.vec.Transformation;
import codechicken.lib.vec.Translation;
import codechicken.lib.vec.Vector3;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

@SideOnly(Side.CLIENT)
public class RenderHalo
{
    public static RenderHalo instance = new RenderHalo();

    private static List<LightCache> renderList = new ArrayList<LightCache>();
    private static Vector3 renderEntityPos = new Vector3();
    private static Vector3 vec = new Vector3();

    private static class LightCache implements Comparable<LightCache>
    {
        final BlockCoord pos;
        final int color;
        final Cuboid6 cube;

        public LightCache(int x, int y, int z, int colorIndex, Cuboid6 cube)
        {
            this.pos = new BlockCoord(x, y, z);
            this.color = colorIndex;
            this.cube = cube;
        }

        private double renderDist() {
            return vec.set(pos.x, pos.y, pos.z).sub(renderEntityPos).magSquared();
        }

		@Override
		public int compareTo(LightCache o) {
			double ra = renderDist();
            double rb = o.renderDist();
            return ra == rb ? 0 : ra < rb ? 1 : -1;
		}
    }

    public static void addLight(int x, int y, int z, int color, Cuboid6 box)
    {
        renderList.add(new LightCache(x, y, z, color, box));
    }

    @ForgeSubscribe
    public void onRenderWorldLast(RenderWorldLastEvent event)
    {
        Tessellator tess = Tessellator.instance;
        WorldClient w = event.context.theWorld;
        EntityLivingBase entity = event.context.mc.renderViewEntity;
        renderEntityPos.set(entity.posX, entity.posY+entity.getEyeHeight(), entity.posZ);

        Collections.sort(renderList);

        GL11.glPushMatrix();
        RenderUtils.translateToWorldCoords(event.context.mc.renderViewEntity, event.partialTicks);
        prepareRenderState();

        Iterator<LightCache> it = renderList.iterator();
        int max = Configurator.lightHaloMax < 0 ? renderList.size() : Configurator.lightHaloMax;
        for(int i = 0; i < max && it.hasNext(); i++)
        {
            LightCache cc = it.next();
            renderHalo(tess, w, cc);
        }

        renderList.clear();
        restoreRenderState();
        GL11.glPopMatrix();
    }

    public static void prepareRenderState()
    {
        GL11.glEnable(GL11.GL_BLEND);
        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE);
        GL11.glDisable(GL11.GL_TEXTURE_2D);
        GL11.glDisable(GL11.GL_LIGHTING);
        GL11.glDisable(GL11.GL_CULL_FACE);
        GL11.glDepthMask(false);
        CCRenderState.reset();
        CCRenderState.startDrawing(7);
    }

    public static void restoreRenderState()
    {
        CCRenderState.draw();
        GL11.glDepthMask(true);
        GL11.glColor3f(1, 1, 1);
        GL11.glEnable(GL11.GL_CULL_FACE);
        GL11.glEnable(GL11.GL_LIGHTING);
        GL11.glEnable(GL11.GL_TEXTURE_2D);
        GL11.glBlendFunc(GL11.GL_SRC_ALPHA, GL11.GL_ONE_MINUS_SRC_ALPHA);
        GL11.glDisable(GL11.GL_BLEND);
    }

    private static void renderHalo(Tessellator tess, World world, LightCache cc)
    {
        CCRenderState.setBrightness(world, cc.pos.x, cc.pos.y, cc.pos.z);
        renderHalo(tess, cc.cube, cc.color, new Translation(cc.pos.x, cc.pos.y, cc.pos.z));
    }

    public static void renderHalo(Tessellator tess, Cuboid6 cuboid, int colour, Transformation t)
    {
        tess.setColorRGBA_I(PRColors.VALID_COLORS[colour].rgb, 128);
        RenderUtils.renderBlock(cuboid, 0, t, null, null);
    }
}

File: src/mrtjp/projectred/core/RenderHalo.scala
package mrtjp.projectred.core

import codechicken.lib.render.{BlockRenderer, CCRenderState}
import codechicken.lib.vec._
import cpw.mods.fml.common.eventhandler.SubscribeEvent
import mrtjp.core.color.Colors
import net.minecraft.client.Minecraft
import net.minecraft.world.World
import net.minecraftforge.client.event.RenderWorldLastEvent
import org.lwjgl.opengl.GL11._

object RenderHalo
{
    private var renderList = Vector[LightCache]()
    private val renderEntityPos = new Vector3
    private val vec = new Vector3

    private class LightCache(val pos:BlockCoord, val color:Int, val cube:Cuboid6) extends Ordered[LightCache]
    {
        def this(x:Int, y:Int, z:Int, c:Int, cube:Cuboid6) = this(new BlockCoord(x, y, z), c, cube)

        private def renderDist = vec.set(pos.x, pos.y, pos.z).sub(renderEntityPos).magSquared

        override def compare(o:LightCache) =
        {
            val ra = renderDist
            val rb = o.renderDist
            if (ra == rb) 0 else if (ra < rb) 1 else -1
        }
    }

    def addLight(x:Int, y:Int, z:Int, color:Int, box:Cuboid6)
    {
        renderList :+= new LightCache(x, y, z, color, box)
    }

    @SubscribeEvent
    def onRenderWorldLast(event:RenderWorldLastEvent)
    {
        if (renderList.isEmpty) return
        val w = Minecraft.getMinecraft.theWorld
        val entity = Minecraft.getMinecraft.renderViewEntity
        renderEntityPos.set(entity.posX, entity.posY+entity.getEyeHeight, entity.posZ)

        renderList = renderList.sorted

        glPushMatrix()

        // Adjust translation for camera movement between frames (using camra coordinates for numeric stability).
        // Note: When porting to MC 1.8, might want to use GlStateManager.translate() here instead.
        glTranslated(
             entity.posX-(entity.posX-entity.lastTickPosX)*event.partialTicks-entity.lastTickPosX,
             entity.posY-(entity.posY-entity.lastTickPosY)*event.partialTicks-entity.lastTickPosY,
             entity.posZ-(entity.posZ-entity.lastTickPosZ)*event.partialTicks-entity.lastTickPosZ)

        prepareRenderState()
        val it = renderList.iterator
        val max = if (Configurator.lightHaloMax < 0) renderList.size else Configurator.lightHaloMax

        var i = 0
        while (i < max && it.hasNext)
        {
            val cc = it.next()
            renderHalo(w, cc)
            i += 1
        }

        renderList = Vector()
        restoreRenderState()
        glPopMatrix()
    }

    def prepareRenderState()
    {
        glEnable(GL_BLEND)
        glBlendFunc(GL_SRC_ALPHA, GL_ONE)
        glDisable(GL_TEXTURE_2D)
        glDisable(GL_LIGHTING)
        glDisable(GL_CULL_FACE)
        glDepthMask(false)
        CCRenderState.reset()
        CCRenderState.setDynamic()
        CCRenderState.startDrawing()
    }

    def restoreRenderState()
    {
        CCRenderState.draw()
        glDepthMask(true)
        glColor3f(1, 1, 1)
        glEnable(GL_CULL_FACE)
        glEnable(GL_LIGHTING)
        glEnable(GL_TEXTURE_2D)
        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA)
        glDisable(GL_BLEND)
    }

    private def renderHalo(world:World, cc:LightCache)
    {
        CCRenderState.setBrightness(world, cc.pos.x, cc.pos.y, cc.pos.z)
        // Make sure to use camera coordinates for the halo transformation.
        val entity = Minecraft.getMinecraft.renderViewEntity
        renderHalo(cc.cube, cc.color,
            new Translation(cc.pos.x-entity.posX, cc.pos.y-entity.posY, cc.pos.z-entity.posZ))
    }

    def renderHalo(cuboid:Cuboid6, colour:Int, t:Transformation)
    {
        CCRenderState.reset()
        CCRenderState.setPipeline(t)
        CCRenderState.baseColour = Colors(colour).rgba
        CCRenderState.alphaOverride = 128
        BlockRenderer.renderCuboid(cuboid, 0)
    }
}